node/core/candidate-validation/src/lib.rs

```diff
@@ -621,28 +625,21 @@ impl ValidationBackend for ValidationHost {
 		timeout: Duration,
 		params: ValidationParams,
 	) -> Result<WasmValidationResult, ValidationError> {
-		let (tx, rx) = oneshot::channel();
-		if let Err(err) = self
-			.execute_pvf(
-				Pvf::from_code(raw_validation_code),
-				timeout,
-				params.encode(),
-				polkadot_node_core_pvf::Priority::Normal,
-				tx,
-			)
-			.await
-		{
-			return Err(ValidationError::InternalError(format!(
-				"cannot send pvf to the validation host: {:?}",
-				err
-			)))
-		}
+		let pvf = Pvf::from_code(raw_validation_code);
 
-		let validation_result = rx
-			.await
-			.map_err(|_| ValidationError::InternalError("validation was cancelled".into()))?;
+		let validation_result = execute_pvf_once(self, pvf.clone(), timeout, params.encode()).await;
```

Comment:
Nit: `params` in their encoded form may be used twice so probably it would be better to encode them once into a variable and use it twice, rather than encoding them twice? I mean cloning the already encoded form is cheaper than encoding the source two times.

Reply:
Given there's a 1 second delay, IMO this won't make a difference

Reply:
Nice catch. I thought about this; so I expect the retry case to be pretty rare. I would rather encode a second time in the retry case, than unconditionally clone and make an extra heap allocation in every case.